### PR TITLE
Revamp Auto Completion Engine on top of the Semantic API

### LINKED ISSUE #26046: Migrate completion engine to the new Semantic API
**Description:**
Currently the completion engine relies on the older AST and BSymbol based APIs. We need to migrate to the new Semantic API and get rid of the older APIs.

### PR DESCRIPTION
**Pull request by nadeeshaan**
Fixes #26046

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/ImplicitNewExpressionNodeContext.java [hunk 1]
```diff
@@ -84,47 +83,47 @@ public ImplicitNewExpressionNodeContext() {
...
                 return Optional.empty();
             }
-            scopeEntry = ((BPackageSymbol) pkgSymbol.get().symbol).scope.entries.entrySet().stream()
-                    .filter(entry -> entry.getKey().value.equals(nameReferenceNode.identifier().text()))
-                    .map(Map.Entry::getValue)
-                    .findAny()
-                    .orElse(null);
+            nameReferenceSymbol = pkgSymbol.get().allSymbols().stream()
+                    .filter(symbol -> symbol.name().equals(nameReferenceNode.identifier().text()))
+                    .findAny();
```
Comment (rasika) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/ImplicitNewExpressionNodeContext.java:98]: `findAny()` will add some randomness into selection criteria even though we expect only one matching symbol. Shall we use `findFirst()` for the sake of consistent in the event of multiple results.

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/ImplicitNewExpressionNodeContext.java [hunk 2]
```diff
@@ -84,47 +83,47 @@ public ImplicitNewExpressionNodeContext() {
...
         } else if (typeDescriptor.kind() == SyntaxKind.SIMPLE_NAME_REFERENCE) {
             SimpleNameReferenceNode nameReferenceNode = (SimpleNameReferenceNode) typeDescriptor;
-            scopeEntry = visibleSymbols.stream()
-                    .filter(entry -> entry.symbol.name.value.equals(nameReferenceNode.name().text()))
-                    .findAny()
-                    .orElse(null);
+            nameReferenceSymbol = visibleSymbols.stream()
+                    .filter(symbol -> symbol.name().equals(nameReferenceNode.name().text()))
+                    .findAny();
```
Comment (rasika) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/ImplicitNewExpressionNodeContext.java:103]: `findAny()` will add some randomness into selection criteria even though we expect only one matching symbol. Shall we use `findFirst()` for the sake of consistent in the event of multiple results.

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/LSAnnotationCache.java [hunk 3]
```diff
@@ -54,26 +57,27 @@
...
-    private static HashMap<PackageID, List<BAnnotationSymbol>> listenerAnnotations = new HashMap<>();
-    private static HashMap<PackageID, List<BAnnotationSymbol>> annotationAnnotations = new HashMap<>();
-    private static HashMap<PackageID, List<BAnnotationSymbol>> externalAnnotations = new HashMap<>();
-    private static HashMap<PackageID, List<BAnnotationSymbol>> varAnnotations = new HashMap<>();
-    private static HashMap<PackageID, List<BAnnotationSymbol>> constAnnotations = new HashMap<>();
-    private static HashMap<PackageID, List<BAnnotationSymbol>> workerAnnotations = new HashMap<>();
-    private static HashMap<PackageID, List<BAnnotationSymbol>> fieldAnnotations = new HashMap<>();
-    private static HashMap<PackageID, List<BAnnotationSymbol>> recordFieldAnnotations = new HashMap<>();
-    private static HashMap<PackageID, List<BAnnotationSymbol>> objectFieldAnnotations = new HashMap<>();
+    private static final HashMap<ModuleID, List<AnnotationSymbol>> typeAnnotations = new HashMap<>();
```
Comment (rasika) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/LSAnnotationCache.java:60]: Shall we use `Map` interface here? `Map<ModuleID, List<AnnotationSymbol>> typeAnnotations = new HashMap<>()`

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/LSAnnotationCache.java [hunk 4]
```diff
@@ -184,13 +152,13 @@ private static void addAttachment(BAnnotationSymbol bAnnotationSymbol,
...
             case OBJECT_FIELD:
-                HashMap<PackageID, List<BAnnotationSymbol>> objFieldAnnotations = new HashMap<>();
+                HashMap<ModuleID, List<AnnotationSymbol>> objFieldAnnotations = new HashMap<>();
                 objFieldAnnotations.putAll(fieldAnnotations);
                 objFieldAnnotations.putAll(objectFieldAnnotations);
                 return objFieldAnnotations;
             case RECORD_FIELD:
             case RECORD_FIELD_WITH_DEFAULT_VALUE:
-                HashMap<PackageID, List<BAnnotationSymbol>> recFieldAnnotations = new HashMap<>();
+                HashMap<ModuleID, List<AnnotationSymbol>> recFieldAnnotations = new HashMap<>();
```
Comment (rasika) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/LSAnnotationCache.java:161]: Shall we use `Map` interface here?

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/ReceiveActionNodeContext.java [hunk 5]
```diff
@@ -46,13 +46,11 @@ public ReceiveActionNodeContext() {
     @Override
     public List<LSCompletionItem> getCompletions(LSContext context, ReceiveActionNode node)
             throws LSCompletionException {
-        // TODO: Following logic can be generalized
-        ArrayList<Scope.ScopeEntry> visibleSymbols = new ArrayList<>(context.get(CommonKeys.VISIBLE_SYMBOLS_KEY));
-        List<Scope.ScopeEntry> filteredWorkers = visibleSymbols.stream()
-                .filter(scopeEntry -> (scopeEntry.symbol.flags & Flags.WORKER) == Flags.WORKER)
+        ArrayList<Symbol> visibleSymbols = new ArrayList<>(context.get(CommonKeys.VISIBLE_SYMBOLS_KEY));
```
Comment (rasika) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/ReceiveActionNodeContext.java:49]: Shall we use `List` interface here?

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/RecordFieldWithDefaultValueNodeContext.java [hunk 6]
```diff
@@ -67,11 +67,8 @@ public RecordFieldWithDefaultValueNodeContext() {
             (2) [module:]TypeName c = module:a<cursor>
              */
             QualifiedNameReferenceNode qNameRef = (QualifiedNameReferenceNode) node.expression();
-            Predicate<Scope.ScopeEntry> filter = scopeEntry -> {
-                BSymbol symbol = scopeEntry.symbol;
-                return symbol instanceof BVarSymbol && (symbol.flags & Flags.PUBLIC) == Flags.PUBLIC;
-            };
-            List<Scope.ScopeEntry> moduleContent = QNameReferenceUtil.getModuleContent(context, qNameRef, filter);
+            Predicate<Symbol> filter = symbol -> symbol instanceof VariableSymbol;
```
Comment (rasika) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/RecordFieldWithDefaultValueNodeContext.java:70]: Can move into a more meaningfull name like `variableFilter`. Can't we use symbolKind over `instaceOf` here?

**Review by rasika — CHANGES_REQUESTED** [review]
Reviewed

File: compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/EnvironmentResolver.java [hunk 7]
```diff
@@ -156,19 +189,19 @@
 
 /**
  * Visitor to lookup symbols based on the location.
- * 
+ *
  * @since 2.0.0
  */
 public class EnvironmentResolver extends BLangNodeVisitor {
```
[the remark this answers is not in the document]
Comment (nadeeshaan) [reply compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/EnvironmentResolver.java:195]: Yes, Compiler API internally depends on it.

**Review by rasika — APPROVED** [review]

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/ImplicitNewExpressionNodeContext.java (hunk 2 above)
Comment (nadeeshaan) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/ImplicitNewExpressionNodeContext.java]: fixed

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/LSAnnotationCache.java (hunk 3 above)
Comment (nadeeshaan) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/LSAnnotationCache.java]: fixed

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/ReceiveActionNodeContext.java (hunk 5 above)
Comment (nadeeshaan) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/ReceiveActionNodeContext.java]: fixed

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/RecordFieldWithDefaultValueNodeContext.java (hunk 6 above)
Comment (nadeeshaan) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/RecordFieldWithDefaultValueNodeContext.java]: fixed

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/LSAnnotationCache.java (hunk 4 above)
Comment (nadeeshaan) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/LSAnnotationCache.java:161]: fixed

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/ErrorTypeParamsNodeContext.java [hunk 8]
```diff
@@ -58,22 +60,26 @@ public ErrorTypeParamsNodeContext() {
...
+            }
+            BallerinaTypeDescriptor typeDesc = ((TypeSymbol) symbol).typeDescriptor();
+            return (typeDesc.kind() == TypeDescKind.MAP || typeDesc.kind() == TypeDescKind.RECORD);
         };
-        List<Scope.ScopeEntry> mappingTypes;
+        List<Symbol> mappingTypes;
         if (this.onQualifiedNameIdentifier(context, nodeAtCursor)) {
-            mappingTypes = QNameReferenceUtil.getModuleContent(context, (QualifiedNameReferenceNode) nodeAtCursor,
-                    predicate.and(scopeEntry -> (scopeEntry.symbol.flags & Flags.PUBLIC) == Flags.PUBLIC));
+            mappingTypes= QNameReferenceUtil.getModuleContent(context,
```
[the remark this answers is not in the document]
Comment (nadeeshaan) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/ErrorTypeParamsNodeContext.java]: fixed

File: compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/EnvironmentResolver.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (rasika) [reply compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/EnvironmentResolver.java:195]: Oh this is inside compiler-api

**Review by rasika — APPROVED** [review]

**Pull request merged**